src/hive-keychain-ceramic.connector.ts

```diff
@@ -0,0 +1,176 @@
+
+import { getPermission } from './get-permission.function'
+import { HiveSignResponse } from './types'
+
+function normalizeAuthSecret(authSecret64: Uint8Array) {
+  const authSecret = new Uint8Array(32)
+  for (let i = 0; i < authSecret.length; i++) {
+    authSecret[i] = authSecret64[i]
+  }
+  return authSecret
+}
+
+export class HiveKeychainCeramicConnector {
+  provider: ThreeIdProvider | undefined
+  DHive: DHiveClient
+  ceramic: CeramicClient
+  idx: IDX
+  loggedIn = false
+
+  constructor(ceramicHost: string = CERAMIC_HOST, hiveHosts: string[] = HIVE_HOSTS) {
+    this.DHive = new DHiveClient(hiveHosts)
+    this.ceramic = new CeramicClient(ceramicHost)
+    this.idx = new IDX({ ceramic: this.ceramic, autopin: true })
+  }
+
+  get cachedSecret(): Uint8Array | null {
+    const secretb64 = localStorage.getItem('hive.ceramic.secret')
+    if (!secretb64) return null
+
+    const buffer = Buffer.from(secretb64, 'base64')
+    return new Uint8Array(buffer)
+  }
+
+  public async login(): Promise<CeramicClient> {
+    if (this.cachedSecret) {
+      // First try to use secret kept in local storage
+      const did = await this.createAndLoginDid(this.cachedSecret, this.ceramic)
+      await this.ceramic.setDID(did)
+    } else {
+      // Otherwise get secret from hive keychain
+      const did = await this.loginWithKeychain(this.ceramic)
+      await this.ceramic.setDID(did)
+    }
+
+    this.loggedIn = true
+    return this.ceramic
+  }
+
+  private setAuthState(authId: string, authSecret: Uint8Array) {
+    localStorage.setItem('hive.ceramic.id', authId)
+    const buffer = Buffer.from(authSecret)
+
+    localStorage.setItem('hive.ceramic.secret', buffer.toString('base64'))
+  }
+
+  private async createAndLoginDid(secret: Uint8Array, ceramic: CeramicClient): Promise<DID> {
+    const authMethod = 'hiveKeychain'
+    //     const provider = new Ed25519Provider(secret)
+    const threeId = await ThreeIdProvider.create({
+      getPermission: getPermission,
+      authSecret: secret,
+      authId: authMethod,
+      ceramic,
+    })
+
+    const provider = threeId.getDidProvider()
+
+    const resolver = ThreeIdResolver.getResolver(ceramic)
+    const did = new DID({ provider, resolver })
+    try {
+      await did.authenticate()
+    } catch (err) {
+      console.error(`Error authenticating DID!`, err)
+    }
+
+    console.log(`did id is ${did.id}`)
+    return did
+  }
+
+  /**
+   * Login with hive keychain
+   * For docs, see https://github.com/stoodkev/hive-keychain/blob/master/README.md
+   */
+  private async loginWithKeychain(ceramic: CeramicClient): Promise<DID> {
+    if (!window.hive_keychain) {
+      throw new Error(`Hive keychain not available!`)
+    }
+
+    const loginResult = await new Promise<HiveSignResponse>((resolve, reject) => {
+      window.hive_keychain?.requestSignBuffer(
+        undefined,
+        'Allow this application to control the Ceramic account associated with your Hive username',
```

Comment:
Can we stick with the universal signing phrase? (note I am in conversations with people about changing this phrase/allowing for more than one phrase)